drivers/auxdisplay/auxdisplay_serlcd.c

```diff
@@ -0,0 +1,361 @@
+	SERLCD_SPECIAL_SET_DD_RAM_ADDRESS = 0x80,
+};
+
+static int auxdisplay_serlcd_send_command(const struct device *dev,
+					  const enum auxdisplay_serlcd_command command)
+{
+	const struct auxdisplay_serlcd_config *config = dev->config;
+	const uint8_t buffer[2] = {SERLCD_BEGIN_COMMAND, command};
+
+	int rc = i2c_write_dt(&config->bus, buffer, sizeof(buffer));
+
+	/*
+	 * commands need some time to process
+	 */
+	k_sleep(K_MSEC(10));
+	return rc;
+}
+
+static int
+auxdisplay_serlcd_send_special_command(const struct device *dev,
+				       const enum auxdisplay_serlcd_special_command command)
+{
+	const struct auxdisplay_serlcd_config *config = dev->config;
+	const uint8_t buffer[2] = {SERLCD_BEGIN_SPECIAL_COMMAND, command};
+
+	int rc = i2c_write_dt(&config->bus, buffer, sizeof(buffer));
+
+	/*
+	 * special commands need some time to process
+	 */
+	k_sleep(K_MSEC(50));
+	return rc;
+}
+
+static int auxdisplay_serlcd_display_on(const struct device *dev)
+{
+	struct auxdisplay_serlcd_data *data = dev->data;
+
+	data->display_state.power = true;
+
+	return auxdisplay_serlcd_send_special_command(dev, SERLCD_SPECIAL_DISPLAY_CONTROL |
+								   data->display_state.byte);
+}
+
+static int auxdisplay_serlcd_display_off(const struct device *dev)
+{
+	struct auxdisplay_serlcd_data *data = dev->data;
+
+	data->display_state.power = false;
+
+	return auxdisplay_serlcd_send_special_command(dev, SERLCD_SPECIAL_DISPLAY_CONTROL |
+								   data->display_state.byte);
+}
+
+static int auxdisplay_serlcd_cursor_set_enabled(const struct device *dev, bool enable)
+{
+	struct auxdisplay_serlcd_data *data = dev->data;
+
+	data->display_state.cursor = enable;
+
+	return auxdisplay_serlcd_send_special_command(dev, SERLCD_SPECIAL_DISPLAY_CONTROL |
+								   data->display_state.byte);
+}
+
+static int auxdisplay_serlcd_position_blinking_set_enabled(const struct device *dev, bool enable)
+{
+	struct auxdisplay_serlcd_data *data = dev->data;
+
+	data->display_state.blinking = enable;
+
+	return auxdisplay_serlcd_send_special_command(dev, SERLCD_SPECIAL_DISPLAY_CONTROL |
+								   data->display_state.byte);
+}
+
+static int auxdisplay_serlcd_cursor_position_set(const struct device *dev,
+						 enum auxdisplay_position type, int16_t x,
+						 int16_t y)
+{
+	static const uint8_t row_offsets[] = {0x00, 0x40, 0x14, 0x54};
+
+	const struct auxdisplay_serlcd_config *config = dev->config;
+	const struct auxdisplay_capabilities capabilities = config->capabilities;
+	const uint16_t columns = capabilities.columns;
+	const uint16_t rows = capabilities.rows;
+	struct auxdisplay_serlcd_data *data = dev->data;
+
+	if (type == AUXDISPLAY_POSITION_ABSOLUTE) {
+		/*
+		 * shortcut for (0,0) position
+		 */
+		if (x == 0 && y == 0) {
+			data->cursor_x = x;
+			data->cursor_y = y;
+			return auxdisplay_serlcd_send_special_command(dev,
+								      SERLCD_SPECIAL_RETURN_HOME);
+		}
+
+		const uint8_t column = MIN(x, columns);
+		const uint8_t row = MIN(y, rows);
```

Comment:
Would expect this to do error checking i.e. returns an error if someone tries to move to the 98th column on a display with 12 columns